### PR TITLE
Try to speed up travis runs.

### PR DESCRIPTION
**Pull request by AlCutter**
- cache test deps by `go install`ing them
- only run tests individually where we require coverage reports.

### CODE REVIEW

File: scripts/presubmit.sh [hunk 1]
```diff
@@ -97,6 +97,11 @@ main() {
     go build ${goflags} ./...
 
     echo 'running go test'
+    # Install test deps so that individual test runs below can reuse them.
+    echo 'getting test deps'
+    go get -t
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Without "./...", this will only do so for the "trillian" package.

Comment (AlCutter) [reply scripts/presubmit.sh]: Done

**Comment by AlCutter** [comment]
PTAL

File: scripts/presubmit.sh [hunk 2]
```diff
@@ -128,8 +132,14 @@ main() {
           ${goflags} "$d"
     done | xargs -I '{}' -P ${GO_TEST_PARALLELISM:-10} bash -c '{}'
 
-    [[ ${coverage} -eq 1 ]] && \
-      cat /tmp/trillian_profile/*.out > /tmp/coverage.txt
+    cat /tmp/trillian_profile/*.out > /tmp/coverage.txt
+  fi
+  else
+      echo go test \
```
Comment (RJPercival) [inline scripts/presubmit.sh]: While this will indeed be fast, that's because it won't actually run the tests. Need to delete the "echo" at the start.

Comment (AlCutter) [reply scripts/presubmit.sh]: hehe, whoops!

File: scripts/presubmit.sh [hunk 3]
```diff
@@ -128,8 +132,14 @@ main() {
           ${goflags} "$d"
     done | xargs -I '{}' -P ${GO_TEST_PARALLELISM:-10} bash -c '{}'
 
-    [[ ${coverage} -eq 1 ]] && \
-      cat /tmp/trillian_profile/*.out > /tmp/coverage.txt
+    cat /tmp/trillian_profile/*.out > /tmp/coverage.txt
+  fi
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Shouldn't be a `fi` right before an `else`.

File: scripts/presubmit.sh [hunk 4]
```diff
@@ -128,8 +132,14 @@ main() {
...
+  else
+      go test \
+        -short \
+        -timeout=${GO_TEST_TIMEOUT:-5m} \
+        ${coverflags} \
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Could remove `${coverflags}`, since it'll always be empty.

File: scripts/presubmit.sh [hunk 5]
```diff
@@ -97,27 +97,31 @@ main() {
...
     # Individual package profiles are written to "$profile.out" files under
     # /tmp/trillian_profile.
     # An aggregate profile is created at /tmp/coverage.txt.
     mkdir -p /tmp/trillian_profile
     rm -f /tmp/trillian_profile/*
 
     for d in $(go list ./...); do
-      # Create a different -coverprofile for each test (if enabled)
+      # Create a different -coverprofile for each test
       local coverflags=
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Can delete this line, and just add `local` to line 124.

File: scripts/presubmit.sh [hunk 6]
```diff
@@ -97,27 +97,31 @@ main() {
     go build ${goflags} ./...
 
     echo 'running go test'
+    # Install test deps so that individual test runs below can reuse them.
+    echo 'getting test deps'
+    go get -t
+    echo 'installed test deps'
+    go test -i ./...
 
+    if [[ ${coverage} -eq 1 ]]; then
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Indentation is a bit funky (though it does make the diff a bit clearer).

File: scripts/presubmit.sh [hunk 7]
```diff
@@ -97,27 +97,31 @@ main() {
     go build ${goflags} ./...
 
     echo 'running go test'
+    # Install test deps so that individual test runs below can reuse them.
+    echo 'getting test deps'
+    go get -t
```
Comment (RJPercival) [inline scripts/presubmit.sh]: Need "./..." for `go get -t` as well as `go test -i`.

File: scripts/presubmit.sh (hunk 4 above)
Comment (AlCutter) [reply scripts/presubmit.sh]: done

File: scripts/presubmit.sh (hunk 3 above)
Comment (AlCutter) [reply scripts/presubmit.sh]: whoops, done

File: scripts/presubmit.sh (hunk 5 above)
Comment (AlCutter) [reply scripts/presubmit.sh]: done

File: scripts/presubmit.sh (hunk 7 above)
Comment (AlCutter) [reply scripts/presubmit.sh]: I think the go get ./... is actually already done in .travis.yml, so I've removed it.

File: scripts/presubmit.sh (hunk 6 above)
Comment (AlCutter) [reply scripts/presubmit.sh]: Fixed, I think

File: scripts/presubmit.sh [hunk 8]
```diff
@@ -97,17 +97,19 @@ main() {
...
-    # An aggregate profile is created at /tmp/coverage.txt.
-    mkdir -p /tmp/trillian_profile
-    rm -f /tmp/trillian_profile/*
-
-    for d in $(go list ./...); do
-      # Create a different -coverprofile for each test (if enabled)
-      local coverflags=
-      if [[ ${coverage} -eq 1 ]]; then
+    # Install test deps so that individual test runs below can reuse them.
+    echo 'installed test deps'
```
Comment (RJPercival) [inline scripts/presubmit.sh]: s/installed/installing/

Comment (AlCutter) [reply scripts/presubmit.sh]: done

**Pull request merged**